sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java

```diff
@@ -442,26 +481,95 @@ public Read usingStandardSql() {
      */
     @Experimental(Experimental.Kind.SOURCE_SINK)
     public Read withTemplateCompatibility() {
-      return toBuilder().setWithTemplateCompatibility(true).build();
+      return new Read(this.inner.withTemplateCompatibility());
+    }
+  }
+
+  /////////////////////////////////////////////////////////////////////////////
+
+  /**
+   * Implementation of {@link BigQueryIO#readRecords(SerializableFunction)} )}.
+   */
+  @AutoValue
+  public abstract static class ReadGenericRecords<T>
+      extends PTransform<PBegin, PCollection<T>> {
+    abstract Builder<T> toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder<T> {
+      abstract Builder<T> setJsonTableRef(ValueProvider<String> jsonTableRef);
+      abstract Builder<T> setQuery(ValueProvider<String> query);
+      abstract Builder<T> setValidate(boolean validate);
+      abstract Builder<T> setFlattenResults(Boolean flattenResults);
+      abstract Builder<T> setUseLegacySql(Boolean useLegacySql);
+      abstract Builder<T> setWithTemplateCompatibility(Boolean useTemplateCompatibility);
+      abstract Builder<T> setBigQueryServices(BigQueryServices bigQueryServices);
+      abstract ReadGenericRecords<T> build();
+
+      abstract Builder<T> setParseFn(
+          SerializableFunction<SchemaAndRecord, T> parseFn);
+      abstract Builder<T> setCoder(Coder<T> coder);
     }
 
+    @Nullable abstract ValueProvider<String> getJsonTableRef();
+    @Nullable abstract ValueProvider<String> getQuery();
+    abstract boolean getValidate();
+    @Nullable abstract Boolean getFlattenResults();
+    @Nullable abstract Boolean getUseLegacySql();
+
+    abstract Boolean getWithTemplateCompatibility();
+
+    abstract BigQueryServices getBigQueryServices();
+
+    abstract SerializableFunction<SchemaAndRecord, T> getParseFn();
+
+    @Nullable abstract Coder<T> getCoder();
+
     @VisibleForTesting
-    Read withTestServices(BigQueryServices testServices) {
-      return toBuilder().setBigQueryServices(testServices).build();
+    Coder<T> inferCoder(CoderRegistry coderRegistry) {
+      if (getCoder() != null) {
+        return getCoder();
+      }
+
+      TypeDescriptor<T> descriptor = TypeDescriptors.outputOf(getParseFn());
+
+      String message =
+          "Unable to infer coder for output of parseFn. Specify it explicitly using withCoder().";
+      checkArgument(descriptor != null, message);
```

Comment:
This was copied from AvroIO but it's buggy there - actually descriptor can never be null, you can omit this check.

Reply:
The function is annotated with `@Nullable`, is that a mistake?

Reply:
Hm which code path are you referring to? I don't see how that can happen, looking at the code currently. It delegates to extractFromTypeParameters which has only 1 return statement, and it returns a type argument of a parameterized type, which can't be null.

Reply:
I see `outputOf` -> `TypeDescriptors.extractFromTypeParameters(type, supertype, extractor)` which has:

```
if (res.hasUnresolvedParameters()) {
      return null;
    } else {
      return (TypeDescriptor<V>) res;
    }
```

https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/values/TypeDescriptors.java#L344

Reply:
Oh, sorry, you're right. I was looking at code in my IDE, looking at a branch where I had changed this, and the branch had been around for so long that I forgot whether this part was submitted or not :-|